metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -1071,7 +1136,11 @@ public ControllerResult<Void> generateRecordsAndResult() throws Exception {
                 log.info("The metadata log appears to be empty. Appending {} bootstrap record(s) " +
                         "at metadata.version {} from {}.", bootstrapMetadata.records().size(),
                         bootstrapMetadata.metadataVersion(), bootstrapMetadata.source());
+                records.add(new ApiMessageAndVersion(
+                        new BeginTransactionRecord().setName("initial transaction"), (short) 0));
                 records.addAll(bootstrapMetadata.records());
+                records.add(new ApiMessageAndVersion(
+                        new EndTransactionRecord(), (short) 0));
```

Comment:
Why do we need the bootstrap records wrapped in a transaction? Aren't they written as an atomic batch?

Reply:
it's not needed but I wanted to get some testing on this. thinking more, we should only begin a transaction here when we're doing upgrade-from-zk and in premigration (but we need to have that code in first...)